articles/intraprocess_communication.md

```diff
@@ -0,0 +1,494 @@
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Introduction
+
+The subscriptions and publications mechanisms in ROS2 fall in two categories:
+
+* intra-process: messages are sent from a publication to subscription via in-process memory.
+* inter-process: messages are sent via the underlying ROS2 middleware layer.
+The specifics of how this happens depend on the chosen middleware implementation and may involve serialization steps.
+
+This design document presents a new implementation for the intra-process communication.
+
+## Motivations for a new implementation
+
+Even if ROS2 supports intra-process communication, the implementation of this mechanism has still much space for improvement.
+Until ROS2 Crystal, major performance issues and the lack of support for shared pointer messages were preventing the use of this feature in real applications.
+
+With the ROS2 Dashing release, most of these issues have been addressed and the intra-process communication behavior has improved greatly ([see ticket](https://github.com/ros2/ros2/issues/649)).
+
+The current implementation is based on the creation of a ring buffer for each `Publisher` and on the publication of meta-messages through the middleware layer.
+When a `Publisher` has to publish intra-process, it will pass the message to the `IntraProcessManager`.
+Here the message will be stored in the ring buffer associated with the `Publisher`.
+In order to extract a message from the `IntraProcessManager` two pieces of information are needed: the id of the `Publisher` (in order to select the correct ring buffer) and the position of the message within its ring buffer.
+A meta-message with this information is created and sent through the ROS2 middleware to all the `Subscription`s, which can then retrieve the original message from the `IntraProcessManager`.
+
+![Current IPC Block Diagram](../img/intraprocess_communication/old_ipc.png)
+
+Several shortcomings of the current implementation are listed below.
+
+### Incomplete Quality of Service support
+
+The current implementation can't be used when the QoS durability value is set to `Transient Local`.
+
+The current implementation does not enforce the depth of the QoS history in a correct way.
+The reason is that there is a single ring buffer per `Publisher` and its size is equal to the depth of the `Publisher`'s history.
+A `Publisher` stores a message in the ring buffer and then it sends a meta-message to allow a `Subscription` to retrieve it.
+The `Subscription` correctly stores meta-messages up to the number indicated by its depth of the history, but, depending on the frequency at which messages are published and callbacks are triggered, it may happen that a meta-message processed from the `Subscription` does not correspond anymore to a valid message in the ring buffer, because it has been already overwritten.
```

Comment:
```
This results in the loss of the message and it is also a difference in behavior between intra and inter-process communication, since, with the latter, the message would have been received.
```